File: config/src/test/scala/net/kemitix/thorp/config/ParseConfigLinesTest.scala
package net.kemitix.thorp.config

import java.nio.file.Paths

import org.scalatest.FunSpec

import scala.jdk.CollectionConverters._

class ParseConfigLinesTest extends FunSpec {

  describe("parse single lines") {
    describe("source") {
      it("should parse") {
        val expected =
          ConfigOptions(
            List[ConfigOption](
              ConfigOption.source(Paths.get("/path/to/source"))).asJava)
        val result = invoke(List("source = /path/to/source"))
        assertResult(expected)(result)
      }
    }
    describe("bucket") {
      it("should parse") {
        val expected =
          ConfigOptions(
            List[ConfigOption](ConfigOption.bucket("bucket-name")).asJava)
        val result = invoke(List("bucket = bucket-name"))
        assertResult(expected)(result)
      }
    }
    describe("prefix") {
      it("should parse") {
        val expected =
          ConfigOptions(
            List[ConfigOption](ConfigOption.prefix("prefix/to/files")).asJava)
        val result = invoke(List("prefix = prefix/to/files"))
        assertResult(expected)(result)
      }
    }
    describe("include") {
      it("should parse") {
        val expected =
          ConfigOptions(
            List[ConfigOption](ConfigOption.include("path/to/include")).asJava)
        val result = invoke(List("include = path/to/include"))
        assertResult(expected)(result)
      }
    }
    describe("exclude") {
      it("should parse") {
        val expected =
          ConfigOptions(
            List[ConfigOption](ConfigOption.exclude("path/to/exclude")).asJava)
        val result = invoke(List("exclude = path/to/exclude"))
        assertResult(expected)(result)
      }
    }
    describe("parallel") {
      describe("when valid") {
        it("should parse") {
          val expected =
            ConfigOptions(List[ConfigOption](ConfigOption.parallel(3)).asJava)
          val result = invoke(List("parallel = 3"))
          assertResult(expected)(result)
        }
      }
      describe("when invalid") {
        it("should ignore") {
          val expected =
            ConfigOptions(List.empty.asJava)
          val result = invoke(List("parallel = invalid"))
          assertResult(expected)(result)
        }
      }
    }
    describe("debug - true") {
      it("should parse") {
        val expected =
          ConfigOptions(List[ConfigOption](ConfigOption.debug()).asJava)
        val result = invoke(List("debug = true"))
        assertResult(expected)(result)
      }
    }
    describe("debug - false") {
      it("should parse") {
        val expected = ConfigOptions.empty
        val result   = invoke(List("debug = false"))
        assertResult(expected)(result)
      }
    }
    describe("comment line") {
      it("should be ignored") {
        val expected = ConfigOptions.empty
        val result   = invoke(List("# ignore me"))
        assertResult(expected)(result)
      }
    }
    describe("unrecognised option") {
      it("should be ignored") {
        val expected = ConfigOptions.empty
        val result   = invoke(List("unsupported = option"))
        assertResult(expected)(result)
      }
    }

    def invoke(lines: List[String]) = {
      (new ParseConfigLines).parseLines(lines.asJava)
    }
  }
}

File: config/src/test/java/net/kemitix/thorp/config/ParseConfigLinesTest.java
package net.kemitix.thorp.config;

import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;

public class ParseConfigLinesTest
        implements WithAssertions {

    private final ParseConfigLines parser = new ParseConfigLines();

    @Test
    @DisplayName("source")
    public void source() {
        testParser("source = /path/to/source",
                ConfigOption.source(Paths.get("/path/to/source")));
    }
    @Test
    @DisplayName("bucket")
    public void bucket() {
        testParser("bucket = bucket-name",
                ConfigOption.bucket("bucket-name"));
    }
    @Test
    @DisplayName("prefix")
    public void prefix() {
        testParser("prefix = prefix/to/files",
                ConfigOption.prefix("prefix/to/files"));
    }
    @Test
    @DisplayName("include")
    public void include() {
        testParser("include = path/to/include",
                ConfigOption.include("path/to/include"));
    }
    @Test
    @DisplayName("exclude")
    public void exclude() {
        testParser("exclude = path/to/exclude",
                ConfigOption.exclude("path/to/exclude"));
    }
    @Test
    @DisplayName("parallel")
    public void parallel() {
        testParser("parallel = 3",
                ConfigOption.parallel(3));
    }
    @Test
    @DisplayName("parallel - invalid")
    public void parallelInvalid() {
        testParserIgnores("parallel = invalid");
    }
    @Test
    @DisplayName("debug - true")
    public void debugTrue() {
        testParser("debug = true",
                ConfigOption.debug());
    }
    @Test
    @DisplayName("debug - false")
    public void debugFalse() {
        testParserIgnores("debug = false");
    }
    @Test
    @DisplayName("comment")
    public void comment() {
        testParserIgnores("# ignore name");
    }
    @Test
    @DisplayName("unrecognised option")
    public void unrecognised() {
        testParserIgnores("unsupported = option");
    }

    public void testParser(String line, ConfigOption configOption) {
        assertThat(invoke(Collections.singletonList(line))).isEqualTo(
                ConfigOptions.create(
                        Collections.singletonList(configOption)));
    }

    public void testParserIgnores(String line) {
        assertThat(invoke(Collections.singletonList(line))).isEqualTo(
                ConfigOptions.create(
                        Collections.emptyList()));
    }

    private ConfigOptions invoke(List<String> lines) {
        return parser.parseLines(lines);
    }
}
